modin/numpy/arr.py

```diff
@@ -0,0 +1,333 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import numpy
+
+from pandas.core.dtypes.common import is_list_like
+from modin.error_message import ErrorMessage
+
+
+class array(object):
+    def __init__(
+        self,
+        object=None,
+        dtype=None,
+        *,
+        copy=True,
+        order="K",
+        subok=False,
+        ndmin=0,
+        like=None,
+        _query_compiler=None,
+        _ndim=None,
+    ):
+        if _query_compiler is not None:
+            self._query_compiler = _query_compiler
+            self._ndim = _ndim
+        elif is_list_like(object) and not is_list_like(object[0]):
+            import modin.pandas as pd
+
+            qc = pd.Series(object)._query_compiler
+            self._query_compiler = qc
+            self._ndim = 1
+        else:
+            expected_kwargs = {
+                "dtype": None,
+                "copy": True,
+                "order": "K",
+                "subok": False,
+                "ndmin": 0,
+                "like": None,
+            }
+            rcvd_kwargs = {
+                "dtype": dtype,
+                "copy": copy,
+                "order": order,
+                "subok": subok,
+                "ndmin": ndmin,
+                "like": like,
+            }
+            for key, value in rcvd_kwargs.copy().items():
+                if value == expected_kwargs[key]:
+                    rcvd_kwargs.pop(key)
+            arr = numpy.array(object, **rcvd_kwargs)
+            self._ndim = len(arr.shape)
+            if self._ndim > 2:
+                ErrorMessage.not_implemented(
+                    "NumPy arrays with dimensions higher than 2 are not yet supported."
+                )
+            import modin.pandas as pd
+
+            self._query_compiler = pd.DataFrame(arr)._query_compiler
+
+    def _absolute(
+        self,
+        out=None,
+        where=True,
+        casting="same_kind",
+        order="K",
+        dtype=None,
+        subok=True,
+    ):
+        result = self._query_compiler.abs()
+        return array(_query_compiler=result, _ndim=self._ndim)
+
+    __abs__ = _absolute
+
+    def _binary_op(self, other):
```

Comment:
We also have to discard the original indices, otherwise, the binary op implementation will try to align indices as it still remembers them:
```python
import modin.config as cfg
import modin.pandas as pd

df1 = pd.DataFrame({"a": [1, 2], "b": [3, 4]})
df2 = pd.DataFrame({"b": [10, 10], "a": [100, 100]})

print("Regular array:")
print(df1.to_numpy() * df2.to_numpy())

cfg.ExperimentalNumPyAPI.put(True)
print("Experimental array:")
print(df1.to_numpy() * df2.to_numpy())
```
Output:
```
Regular array:
[[ 10 300]
 [ 20 400]]
Experimental array:
array([[100,  30],
       [200,  40]])
```
I would personally prefer the indices to be discarded by the `query_compiler.binary_op` implementation itself using some kind of `ignore_indices=True` flag passed from here. This way we could easily support a commonly used case `md_series[subset1] * md_series[subset2].values` in the future.

Reply:
Even if we discard the original indices, there will be certain cases that won't work.

```python
import pandas

df1 = pandas.DataFrame({"a": [1, 2], "b": [3, 4]})
df2 = pandas.DataFrame({"c": [1, 2]})
df1.columns = [0, 1]
df2.columns = [0]

df1 + df2
   0   1
0  2 NaN
1  4 NaN

df1.values + df2.values
array([[2, 4],
       [4, 6]])
```

We could convert pandas DataFrames to NumPy arrays in the compute kernels to get proper behavior for modin.numpy. However, I think that really mixes up pandas and numpy storage formats. So I would recommend to start implementing the Modin layers for modin.numpy to make the things work correctly and get the proper Modin structure without mixing up the things.

Reply:
so for now we want basic support and I propose that at least a case like this should work in this branch to proceed:
```python
import modin.pandas as pd

df1 = pd.DataFrame({"a": [1, 2, 3, 4], "b": [3, 4, 5, 6], "c": [10, 20, 30, 40]})
df2 = pd.DataFrame({"x": [11, 12, 13, 14], "y": [15, 4, 5, 6], "z": [4, 3, 2, 1]})

indexer1 = pd.Series([1, 0, 2]).to_numpy()
indexer2 = pd.Series([3, 2, 1]).to_numpy()

# First and second '*' operands have completely different indices but '.to_numpy()' should ignore them
res = df1[["a", "c"]].to_numpy()[indexer1] * df2[["x", "y"]].to_numpy()[indexer2]
print(res)
```

Reply:
Fixed!